# Mental Health in Tech Industry Analysis 2014-2019
### *author: Celine Ng*
### *April 2024*

# Table of Contents 
1. Notebook Preparation
    1.1 Libraries Installation
    1.2 Data Retrieval
    1.3 Initial Data Cleaning
2. Dataset 'Answer'

# 1. Notebook preparation
## 1.1 Libraries Installation

In [4]:
!pip install pandas
!pip install matplotlib
!pip install seaborn

In [5]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import sqlite3 as sql

## 1.2 Data Retrieval
Objective: Preview datasets and have a general understanding of each of the 
tables.

Create a relative path to the SQLite database and connect to the SQLite 
database

In [6]:
database_path = 'mental_health.sqlite'
conn = sql.connect(database_path)

Check how many tables is in the database and the names of them

In [7]:
query = 'SELECT name FROM sqlite_master WHERE type="table";'
tables = pd.read_sql_query(query, conn)
print(tables)

       name
0    Answer
1  Question
2    Survey


**Answer**

In [10]:
answer_df = pd.read_sql('SELECT * FROM Answer', conn)
answer_df.head()

AnswerText  SurveyID  UserID  QuestionID
0         37      2014       1           1
1         44      2014       2           1
2         32      2014       3           1
3         31      2014       4           1
4         31      2014       5           1

In [23]:
answer_df.shape

(236898, 4)

In [24]:
answer_df.UserID.nunique()

4218

There are 236898 columns in the Answer dataset, with 2 foreign keys, 
SurveyID and QuestionID. 
In total, 4218 people answered the survey.

**Question**

In [14]:
question_df = pd.read_sql('SELECT * FROM Question', conn)
question_df

questiontext  questionid
0                                    What is your age?           1
1                                 What is your gender?           2
2                         What country do you live in?           3
3    If you live in the United States, which state ...           4
4                               Are you self-employed?           5
..                                                 ...         ...
100  Do you think that team members/co-workers woul...         114
101  If yes, what condition(s) have you been diagno...         115
102  If maybe, what condition(s) do you believe you...         116
103  Which of the following best describes your wor...         117
104                              Do you work remotely?         118

[105 rows x 2 columns]

In [25]:
question_df.shape

(105, 2)

In [ ]:
user_questions = [1, 2, 3, 4, 5, 13, 20, 21, 22]
mental_health_questions = [6, 7]
company_questions = [8, 9]
company_mental_health_questions = [10, 11, 15, 16, 23]
opinion_questions = [12, 14, 17, 18, 19, 24]

There are 105 distinct questions. The dataset includes a primary key,
questionID.

**Survey**

In [13]:
survey_df = pd.read_sql('SELECT * FROM Survey', conn)
survey_df

SurveyID                    Description
0      2014  mental health survey for 2014
1      2016  mental health survey for 2016
2      2017  mental health survey for 2017
3      2018  mental health survey for 2018
4      2019  mental health survey for 2019

We have a missing year 2015. This dataset includes a primary key, SurveyID.

## 1.3 Initial Data Cleaning
Objective: Check for missing values and duplicates 

In [17]:
datasets = [answer_df, question_df, survey_df]

1. Check for duplicates in each dataset.

In [18]:
for dataset in datasets:
    print(dataset.duplicated().any())

False
False
False


2. Check for missing values in each dataset.

In [21]:
for dataset in datasets:
    print(dataset.isna().any().any())

False
False
False


There is no missing values or duplicated rows in all 3 datasets.

# 2. Dataset 'Answer'
## 2.1 Groupby Survey

In [26]:
answer_df

AnswerText  SurveyID  UserID  QuestionID
0                       37      2014       1           1
1                       44      2014       2           1
2                       32      2014       3           1
3                       31      2014       4           1
4                       31      2014       5           1
...                    ...       ...     ...         ...
236893               Other      2016    2689         117
236894             Support      2016    2690         117
236895  Back-end Developer      2016    2691         117
236896     DevOps/SysAdmin      2016    2692         117
236897               Other      2016    2693         117

[236898 rows x 4 columns]

In [36]:
query = '''SELECT SurveyID, COUNT(1) AS TotalReplies
 FROM Answer 
 GROUP BY SurveyID;
 ''' 

replies_over_time = pd.read_sql_query(query, conn)
replies_over_time

SurveyID  TotalReplies
0      2014         32760
1      2016         88238
2      2017         57456
3      2018         31692
4      2019         26752

In [38]:
query = '''SELECT *
FROM Answer
ORDER BY SurveyID, QuestionID;
'''

questions_for_each_survey = pd.read_sql_query(query, conn)
questions_for_each_survey

AnswerText  SurveyID  UserID  QuestionID
0              37      2014       1           1
1              44      2014       2           1
2              32      2014       3           1
3              31      2014       4           1
4              31      2014       5           1
...           ...       ...     ...         ...
236893         -1      2019    4214          89
236894      White      2019    4215          89
236895         -1      2019    4216          89
236896         -1      2019    4217          89
236897      White      2019    4218          89

[236898 rows x 4 columns]